### ISSUE TITLE
How to make it not damn sloooooow with PHP application? (mounted volume IO)

### ISSUE THREAD

**Issue by kmajzlik** [docker/for-mac#2903]
- [x] I have tried with the latest version of my channel (Stable or Edge)
  - [x] I have submitted Diagnostics
  - [x] I have included the Diagnostics ID in this Issue
  - macOS Version: 10.13.4

### Expected behavior
Work ~ with usable speed.

### Actual behavior
It is unusable for real development of PHP apps (Drupal 7/8 in my case).

### Information
```
Docker for Mac: version: 18.05.0-ce-rc1-mac63 (7e2bb492d35dc35ec0037d8024b7e7be57fc72a9)
macOS: version 10.13.4 (build: 17E202)
logs: /tmp/18B232F3-6F30-48D8-B2CA-7A2B609403D7/20180513-195513.tar.gz
[OK]     vpnkit
[OK]     virtualization hypervisor
[OK]     vmnetd
[OK]     dns
[OK]     driver.amd64-linux
[OK]     virtualization VT-X
[OK]     app
[OK]     moby
[OK]     system
[OK]     moby-syslog
[OK]     kubernetes
[OK]     files
[OK]     env
[OK]     virtualization kern.hv_support
[OK]     osxfs
[OK]     moby-console
[OK]     logs
[OK]     docker-cli
[OK]     disk
```
diagnosis ID: 18B232F3-6F30-48D8-B2CA-7A2B609403D7

### Steps to reproduce the behavior

  1. try to work anyhow with PHP application like Drupal in mounted volume

It is damn slow in disk operations. Try this simple benchmark https://github.com/martin-klima/benchmark-php
With mounted volume (- ".:/var/www/html:delegated") the benchmark gives me 60 secs for IO ops.
In container without mounted volume, downloaded same benchmark and i am under 1 sec for IO ops! This really is difference.

I spent whole weekend to try several things, i switched from Stable to Edge(to switch qcow to raw), i tried to switch volume's flag (cached/delegated). I tried virtualbox on my mac (ubuntu installed) it is comparable with that Docker without mounted volume. I tried Docker in that virtual linux and there is no performance problem.

I have read many issues/blogs etc about this and didnt found real solution (qcow -> raw is not important). I want to use Docker, i love my Mac but it is unusable together for PHP world (and i have seen many same issues for JS world and node_modules directory).
  
Options which i have not tried:
- docker-sync
- docker machine

**Comment by akimd** [comment]
Hi!
We are aware of this issue, and there is ongoing work to address this.  Maybe @djs55 or @pgayvallet will want to add information about this?
Cheers!

**Comment by kmajzlik** [comment]
We have tried again on more machines exactly same container, conditions and things like that.
If you run that benchmark in mounted volume(which you need for development to see code in your IDE) it is 30-100x slower compared to moved benchmark file to /root (for example, simply out of mounted volume). Macbook PRO 2016/2017 is little bit better(36,2 vs 1,0 sec) than early 2015(57,2 vs 0,751 sec) but still it makes huge performance impact.

**Comment by milkovsky** [comment]
Might be related: https://github.com/docker/compose/issues/3419#issuecomment-221793401
https://maketips.net/tip/461/docker-build-is-slow-or-docker-compose-build-is-slow

**Comment by YRM64** [comment]
Developers,
Narcisa Zysman, Senior Product Manager, CAST says that there are 10 ways to speed up your PHP Applications:

1) "Avoid Needless Copies of your Variables"
     Sometimes, developers think they can make the code cleaner by copying predefined variables to    
     variables with shorter names before working with them.
2) "Avoid String Concatenations in Loops"
      When placed in a loop, string concatenation results in the creation of large numbers of temporary
      objects and unnecessary use of the garbage collector.
3)  "Avoid Function Tests in Loop Conditionals"
       Function tests in loop conditionals decrease application performance.  When you use a loop
       through an array, the function is called every time.  Instead, do a count() beforehand, store the
       value in a variable, and use it for the test.
4)    "Avoid Passing Function Variables by Reference"
        In most cases, functions only need to use the values passed by their parameters, without altering    
        the values of such parameters.
5)    "Avoid Using Relative Paths"
        For Better Performance, it is highly advised to try and minimize the use of relative paths for file 
        inclusion.
6)    "Avoid Methods with Object Instantiation in Loops"
         PHP is an object-oriented language, and one of the fundamental object-oriented performance 
         management principles is to avoid excessive object creation.
7)    "Avoid Using Period for "Echo" Function"
        When using the "echo" function with periods, PHP concatenates the strings before it outputs 
        them, decreasing the performance.  Using commas instead of periods outputs them in the 
        declared order, with no extra processing.
8)     "Avoid Using Regular Expressions"
         Regular expressions are very useful, but also very time-consuming.  Limit their use.
9)      "Avoid Using Double Quotes for Long Strings without Variables"
          For strings declared using double quotes, PHP does extra processing to find potential variables
          which could be used inside those strings.
10)    "Avoid Using Include_Once"
          The structure include_once includes and checks the specified file during the script execution.
          The functionality is similar to include, but in the case of include_once, verification is done to
          check that the file has been included only once.
       
Ladies and Gents, you can access additional postings from Narcisa Zysman at castsoftware.com.  You can also view CAST Highlights and CAST AIP (castsoftware.com).  Other tools are available for speeding up your PHP applications.  This post by Narcisa Zysman is available in full context at www.castsoftware.com.

**Comment by YRM64** [comment]
Developers,

 Narcisa Zysman, Senior Product Manager, CAST says that there are 10 ways to speed up your PHP Applications:

1. "Avoid Needless Copies of your Variables" 
     Sometimes, developers think they can make the code "cleaner" by copying predefined variables to          
     variables with shorter names before working with them.
2. "Avoid String Concatenations in Loops" 
     When placed in a loop, string concatenation results in the creation of large numbers of temporary 
     objects and unnecessary use of the garbage collector.
3. "Avoid Function Tests in Loop Conditionals" 
     Function tests in loop conditionals decrease application performance. When you use a loop 
     through an array, the function is called every time. Instead, do a count() beforehand, store the 
     value in a variable, and use it for the test.
4. "Avoid Passing Function Variables by Reference"
      In most cases, functions only need to use the values passed by their parameters, without altering
     the values of such parameters.
5. "Avoid Using Relative Paths" 
     For Better Performance, it is highly advised to try and minimize the use of relative paths for file 
     inclusion.
6. "Avoid Methods with Object Instantiation in Loops" 
      PHP is an object-oriented language, and one of the fundamental object-oriented performance 
      management principles is to avoid excessive object creation.
7. "Avoid Using Period for "Echo" Function" 
     When using the "echo" function with periods, PHP concatenates the strings before it outputs 
     them, decreasing the performance.  Using commas instead of periods outputs them in the 
     declared order, with no extra processing -- reducing your chances of hurting app performance.
8. "Avoid Using Regular Expressions"
      Regular expressions are very useful, but also very time-consuming.  Limit their use.
 9. "Avoid Using Double Quotes for Long Strings without Variables"
      For strings declared using double quotes, PHP does extra processing to find potential variables    
      which could be used inside those strings.
10.  "Avoid Using Include_Once" 
        The structure include_once includes and checks the specified file during the script execution.     
        The functionality is similar to include, but in the case of include_once, verification is done
        to check that the file has been included only once.

Ladies and Gents, you can access additional postings from Narcisa Zysman at castsoftware.com. You can also view CAST Highlights and CAST AIP (castsoftware.com). Other tools are available for speeding up your PHP applications. This post by Narcisa Zysman is available in full context at www.castsoftware.com.

**Comment by thaJeztah** [comment]
@YRM64 probably well intended, but it's a bit out of scope for this issue

**Comment by akimd** [comment]
Closing in favor of the older issue https://github.com/docker/for-mac/issues/77.

**Comment by krebbi** [comment]
@akimd 
closing this in favor of another issue that was closed for comments on 5 May 2017?

**Comment by kmajzlik** [comment]
There is no solution. You can make it faster when you make vendor (or node_modules in javascript) non-shared volume. And sometime sync from container to host it if you need inspect it in IDE:
`docker cp "$(docker-compose ps -q php)":/var/www/html/vendor ./`

**Comment by radovanobal** [comment]
Experiencing this on a m.2 ssd drive with 32GB RAM, gen 6 i7 CPU. 
It takes 15 seconds to do 6 queries and sort the data. Running this code outside of the container takes miliseconds.
